### DISCUSSION TITLE
How to find out if messages are rendered

### DISCUSSION THREAD

**Discussion by hgfrey** [primefaces/community#1159] [PrimeFaces] [answered]
How can I find out if there are messages on a page? Best after the page has been built because I want to set the focus then? But I can also remember the information in @PostConstruct.

**Comment by melloware** [comment]
Are you asking for client side or server side if messages are rendered?  You mention focus so I assume you mean client side?

**Reply by hgfrey** [reply]
Server Side, I think. My current solution looks like this:

I've added an outputPanel with a deferred loading which contains a remode command:

```
	<p:outputPanel id="after-page-load" deferred="true" deferredMode="load">
		<p:remoteCommand
				id="after-page-load-remote-command"
				global="false"
				autoRun="true"
				action="#{itAktivitaetPageController.afterPageLoad()}"
				ignoreAutoUpdate="true"
				process="@this"
				update="@none"/>
	</p:outputPanel>

```
In the remote command I try to find out where tp place the focus.

```
    public void afterPageLoad() {
        if (this.pageContainsMessages) {
            messagesUtils.handleMessageLink();
        }
    }

```
In handleMessageLink I add the role=button to the link and place the focus to the elment that has the style class 'ui-messages-close'.

And in the postContruct Routine I find out if there is any message rendered:

```
    @PostConstruct
    public void postConstruct() {
        pageContainsMessages = FacesContext.getCurrentInstance().getMessageList().stream().anyMatch(FacesMessage::isRendered);
    }

```
These snippets already do everything I need. But perhaps there is a more elegant and, above all, simpler solution.

**Reply by hgfrey** [reply]
Unfortunately, I found that my solution above doesn't work when I use the ConversationScope and switch between dialogs here. The PostConstruct is not executed.

**Comment by hgfrey** [comment]
I would like to repeat my question again. Do I have a way on the server side to find out if a message is displayed on the current page?

**Reply by tandraschko** [reply]
yes, myfaces does something similar: https://github.com/apache/myfaces/blob/0af66ca29c8982713b8b1b444eb0a0346cb1104d/impl/src/main/java/org/apache/myfaces/lifecycle/RenderResponseExecutor.java#L126

**Reply by hgfrey** [reply]
As I wrote, reading the message list in PostConstruct works. However, this is not executed when I change pages in the ConversationScope or later when drawing the mask, this information is no longer available. Unfortunately, I have to explicitly remember that I have issued a message.

**Reply by tandraschko** [reply]
You propably need a prerender phaselister then

**Reply by hgfrey** [reply]
I think that PreRenderer is a good place where the information is reliably available. I'll check this out. But how does the information get to my backing beans?

**Reply by tandraschko** [reply, marked as answer]
Inject the bean into the phaselistener

**Reply by hgfrey** [reply]
we have hundreds of backing beans :-(

**Reply by tandraschko** [reply]
Then fire a CDI event and observe it in the bean If you need it

**Reply by hgfrey** [reply]
I'll try it.

**Reply by melloware** [reply]
@hgfrey can you respond if you got it working so we can mark a solution to this thread?

**Reply by hgfrey** [reply]
This seems to work. Is it ensured that the listener only calls the beans belonging to the session?

Despite it. Actually, I would have expected that a displayed message would be a UI-component just like any other component. I get the message UI-Component, but it is empty.
